### ISSUE TITLE
UC_HOOK_MEM_* not working in Python bindings

### ISSUE THREAD

**Issue by esanfelix** [unicorn-engine/unicorn#212]
This might be a non-issue, as I just built Unicorn and started playing around with it, so please bear with me if that's the case.

I've been trying to trace memory read/write accesses with the Python bindings, and cannot get it to work. I've modified the sample_x86.py to create a stripped down example:

```python
#!/usr/bin/env python
# Sample code for X86 of Unicorn. Nguyen Anh Quynh <aquynh@gmail.com>
from __future__ import print_function
from unicorn import *
from unicorn.x86_const import *


X86_CODE16 = '\x00\x00' # add   byte ptr [bx + si], al


# callback for tracing memory access (READ or WRITE)
def hook_mem_access(uc, access, address, size, value, user_data):
    if access == UC_MEM_WRITE:
        print(">>> Memory is being WRITE at 0x%x, data size = %u, data value = 0x%x" \
                %(address, size, value))
    else:   # READ
        print(">>> Memory is being READ at 0x%x, data size = %u" \
                %(address, size))

def test_x86_16():
    print("Emulate x86 16-bit code")
    try:
        # Initialize emulator in X86-16bit mode
        mu = Uc(UC_ARCH_X86, UC_MODE_16)
        mu.hook_add(UC_HOOK_MEM_READ | UC_HOOK_MEM_WRITE, hook_mem_access)

        # map 8KB memory for this emulation
        mu.mem_map(0, 8 * 1024)

        # set CPU registers
        mu.reg_write(UC_X86_REG_EAX, 7)
        mu.reg_write(UC_X86_REG_EBX, 5)
        mu.reg_write(UC_X86_REG_ESI, 6)

        # write machine code to be emulated to memory
        mu.mem_write(0, X86_CODE16)

        # emulate machine code in infinite time
        mu.emu_start(0, len(X86_CODE16))

        # now print out some registers
        print(">>> Emulation done. Below is the CPU context")

        tmp = mu.mem_read(11, 1)
        print("[0x%x] = 0x%x" %(11, tmp[0]))

    except UcError as e:
        print("ERROR: %s" % e)


if __name__ == '__main__':
    test_x86_16()
```

What I expect to see is my callback being called for a READ and then a WRITE. What I see is no calls to the callback:
```bash
➜  python  python testMem.py 
Emulate x86 16-bit code
>>> Emulation done. Below is the CPU context
[0xb] = 0x7
```

I've tried the same thing with the sample x86 16-bit code directly in C, and the result is as expected:
```bash
➜  samples git:(master) ✗ ./sample_x86 -16
Emulate x86 16-bit code
>>> Memory is being READ at 0xb, data size = 1
>>> Memory is being WRITE at 0xb, data size = 1, data value = 0x6c
>>> Emulation done. Below is the CPU context
>>> Read 1 bytes from [0xb] = 0x6c
```
Am I doing anything wrong or are the python bindings not functioning properly?

**Comment by gaffe23** [comment]
I'm getting what appears to be correct output from your Python code:

````
Emulate x86 16-bit code
>>> Memory is being READ at 0xb, data size = 1
>>> Memory is being WRITE at 0xb, data size = 1, data value = 0x7
>>> Emulation done. Below is the CPU context
[0xb] = 0x7
````

You might want to try re-building and re-installing the Python bindings.

**Comment by esanfelix** [comment]
Indeed that is the expected output. I've just set up on OSX and get the expected output as well.

I have removed everything, cloned the repository and reinstalled on my Ubuntu machine and I am still having the same issue. It seems more related to my setup than to unicorn then, any ideas on how to best debug it?

**Comment by aquynh** [comment]
i can confirm that your code works as expected. below is the output:

    Emulate x86 16-bit code
    >>> Memory is being READ at 0xb, data size = 1
    >>> Memory is being WRITE at 0xb, data size = 1, data value = 0x7
    >>> Emulation done. Below is the CPU context
    [0xb] = 0x7

you must make sure that you have latest code installed - **together with the latest Python binding**. if the core & the Python binding are not syn, you might have some unknown issues.

so with the latest code pull from Github, do this:

    $ ./make.sh
    $ sudo ./make.sh install
    $ cd bindings/python
    $ sudo make install

i think this would solve your issue.

**Comment by esanfelix** [comment]
Hi  @aquynh ,

Those are exactly the steps I performed. I also tried going back to the 0.9 release version using the tarball from the website with the same result.

The funny thing is that this is happening only in one of 3 machines I tested with. Seems to be related to that specific setup, so I guess we can close the issue. I am still curious as to what is causing this problem on that machine though, might investigate if I have time.
